### PR TITLE
Ast impl

### PR DESCRIPTION
**Pull request by vlad-scherbich**
I'm going to do this iteratively. This PR simply fixes all existing issues with the AST (besides defining `receive`, which I will do shortly). 

Let me know how this looks. I'll start adding new things to this soon.

### CODE REVIEW

**Comment by Howon** [comment]
Squash your commits

**Comment by Howon** [comment]
Why are you submitting parser in this?

File: src/ast.mli [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+type p_type = Int_t | Bool_t | Double_t | Char_t | Unit_t
+
+type cont_type = String_t | List_t | Set_t | Map_t | Tuple_t
+
+type actor_op =
+  | Actor_send
+  | Actor_broadcast
+  | Actor_receive
+
+type types =
```
Comment (anthonyholley) [inline src/ast.mli:20]: screen shot 2016-10-29 at 2 47 11 pm
I think the code needs to be changed to this^

File: src/ast.mli [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  | Primitive of p_type
+  | Container of cont_type
+  | Actor_t of actor_op
+
+type opt_type = Maybe_t of types | Some_t of types | None_t
+
+(* todo: not sure if works *)
+type opt_nested_type =
```
Comment (anthonyholley) [inline src/ast.mli:28]: This is not needed (unless I'm mistaken)

**Review by anthonyholley — COMMENTED** [review]
Changes to ast concerning optionals

File: src/ast.mli (hunk 2 above)
Comment (vlad-scherbich) [reply src/ast.mli:28]: We need a guru here ... @Howon does this sound right?

File: src/ast.mli (hunk 1 above)
Comment (vlad-scherbich) [reply src/ast.mli:20]: Quite possible, I wasn't sure about my implementation of opt. 

I think one thing that's possible with this definition, but shouldn't be allowed is:
`None<int> = ...` ,
Agreed? If so, we'll need to modify accordingly.

**Comment by Howon** [comment]
who the fuck is commenting using my account

**Comment by vlad-scherbich** [comment]
@Howon no one is using your account. Apparently I've edited your comments, instead of replying to them. I don't see a way to reply to them directly.  They've been restored.

Why squash during a PR? Does github not handle this correctly when you merge? 
I can always squash just before merging too ...

**Comment by Howon** [comment]
please squash ur commits to sizable chunks before make a pr

**Comment by vlad-scherbich** [comment]
Oh, these commits were mostly not mine (I only made one commit). I've rebased off another branch for testing, that's why it looks weird. I'll clean up this branch, or maybe start a new PR going to be in a clean state.

File: src/ast.mli (hunk 2 above)
Comment (aethos) [reply src/ast.mli:28]: why do we have optional types again?

**Comment by aethos** [comment]
let's merge thsi

**Comment by Howon** [comment]
your scanner is not upto date
please use our latest scanner

File: src/ast.mli (hunk 1 above)
Comment (anthonyholley) [reply src/ast.mli:20]: 'none<int>' is not possible with this definition as the None_param is just a placeholder for the lack of a parameter following a 'none'. So when an optional is a 'none', the optional_type will be None_t (which represents the 'none' part) and the optional_param will be None_param (which represents the "nothing", or empty string, that follows 'none').

Comment (anthonyholley) [reply src/ast.mli:20]: Actually I'm talking nonsense, for some reason I was assuming that for this particular case the ast would automatically know that only None_param should be selected if the type is none. This is a quick fix though.

File: src/ast.mli (hunk 2 above)
Comment (vlad-scherbich) [reply src/ast.mli:28]: Guys, let's get on the same page about this. We should only spend time trying to get this right if optionals are needed indeed. 

LRM states

```
Oscar supports optional values for NullPointerException-safety.
    maybe<int> perhaps = some<int>(0);
    maybe<int> no = none;
```

Why couldn't we just compare variables to `none` for checking against NULL-ness?

**Pull request closed without merging**